src/Microsoft.Restier.Publishers.OData/Model/RestierModelBuilder.cs

```diff
@@ -49,7 +55,19 @@ class RestierModelBuilder : IModelBuilder
 
             context.EntitySetTypeMapCollection.Clear();
 
+            ExtendModel(builder);
+
             return builder.GetEdmModel();
         }
+
+        /// <summary>
+        /// This method allow user to extend the model with more entity type and entity set before operation model builder starts
+        /// A sub class need to be implemented and register as DI service like "AddTransient&lt;RestierModelBuilder, CustomizedRestierModelBuilder&gt;"
+        /// </summary>
+        /// <param name="builder"></param>
+        public virtual void ExtendModel(ODataConventionModelBuilder builder)
```

Comment:
Just to make sure I point it out, if you put all of your existing awesomeness into the BuildModelAsync() function, it will construct the model for you by default. Anyone needing to customize the process would just call base.BuildModelAsync() in their override code, and then finish with their modifications. That way, people that add functions, actions, or non-entity types to their APIs can do so after the existing model has been built. Also, this method would make it really easy to add things like ignored properties and what-not, without having to take over building the *entire* model, like you have to do right now.

It would look like this:

```cs
///<summary>
/// Handles the automagic process of constructing a ModelBuilder instance from a given DbContext.
///</summary>
public class CustomRestierModelBuilder : RestierModelBuilder 
{
    ///<summary>
    /// Creates a new instance of the CustomRestierModelBuilder.
    ///</summary>
    /// <param name="namespace">The namespace to use for the ModelBuilder.</param>
    public CustomRestierModelBuilder(string namespace) : base(namespace)
    {
    }

    ///<summary>
    /// Please start documenting these methods so that Intellisense works.
    </summary>
    public override async Task BuildModelAsync(ModelContext context, CancellationToken cancellationToken)
    {
        base.BuildModelAsync(context, cancellationToken);

        // My custom stuff here
        ModelBuilder.EntitySet<Contact>("Contacts").EntityType.Ignore(c => c.IsDeleted);
        ModelBuilder.Action("SomeAction").Returns<SomeAwesomeType>();
        ....
    }

}
```
For the end user, this will be **so much cleaner** than the implementation you have now. Super-simple.

Expecting that EF people won't need to touch the model is a bad assumption. Especially if using the attributes to get Actions to route properly is as difficult for everyone else as it was for me this past weekend.

I wouldn't extract stuff into different methods right now. Keep It Simple. Don't over-engineer it. The whole point of the framework is to make building these APIs ridiculously simple. You have to be **ruthless** with the design simplicity, and don't engineer features no one is asking for because few are using it. Right. @markdstafford? :)